Add definition of validity scopes over aggregation parameters. (#161)
cfrg/draft-irtf-cfrg-vdaf: draft-irtf-cfrg-vdaf.md

identifier is equal to the integer in range `[0, SHARES)` that matches the index
of `input_share` in the sequence of input shares output by the Client.

> Concrete DAFs implementations MAY also specify a *validity scope* for

Reviewer: nit: The I-D tools don't render this as bold-face text as you would hope. It will instead leave these asterisks as they are. I would replace with quotations.

```suggestion
Concrete DAFs implementations MAY also specify a "validity scope" for
```

> example, public keys are required. On the other hand, we could consider
> defining this class of schemes as a different primitive. See issue#77.

## Validity of Aggregation Parameters {#sec-vdaf-validity-scopes}

> Similar to DAFs (see {{sec-vdaf-validity-scopes}}), VDAFs MAY impose

Reviewer 2: nit: link points to wrong section
```suggestion
Similar to DAFs (see {{sec-daf-validity-scopes}}), VDAFs MAY impose
```

restrictions for input shares and aggregation parameters. Protocols using a VDAF

> MUST ensure that for each input share and aggregation parameter `agg_param`, The

Reviewer 2: nit: copyedit
```suggestion
MUST ensure that for each input share and aggregation parameter `agg_param`, the
```

~~~
{: #prio3-prep-state title="Preparation state for Prio3."}

### Validity of Aggregation Parameters

Every input share MUST only be used once, regardless of the aggregation
parameters used.

~~~
def is_valid(agg_param, previous_agg_params):

>     return (len(previous_agg_params) == 0)

Reviewer: nit (totally feel free to ignore)
```suggestion
    return len(previous_agg_params) == 0
```

Reviewer: Here and below.

Author: Done.

The algorithms below make use of the auxiliary function `decode_input_share()`
defined in {{poplar1-auxiliary}}.

> 

Reviewer: Given the importance of this text, I think we should lift it to a subsection somewhere. We want to make sure it's not buried

Author: Done.

~~~
{: #poplar1-prep-state title="Preparation state for Poplar1."}

### Validity of Aggregation Parameters

> Aggregation parameters are valid for a given input share, if no aggregation

Reviewer 2: nit: unnecessary comma
```suggestion
Aggregation parameters are valid for a given input share if no aggregation
```

parameter with the same level has been used with the same input share before.
The whole preparation phase MUST NOT be run more than once for a given
combination of input share and level.

~~~
def is_valid(agg_param, previous_agg_params):
    (level, _) = agg_param

>     return len(filter(lambda (other_level, _): other_level == level, \
>                      previous_agg_params)) == 0

Reviewer 2: This may be more idiomatic and clearer.
```suggestion
    return all(
        level != other_level
        for (other_level, _) in previous_agg_params
    )
```

Author: Done.

across multiple batches. It is RECOMMENDED that the nonce generated by the
Client be used by the Aggregators for replay protection.

## Requirements for the Aggregation Parameters

> As described in {{sec-daf-validity-scopes}}, DAFs and VDAFs may impose

Reviewer: ```suggestion
As described in {{sec-daf-validity-scopes}} and {{sec-vdaf-validity-scopes}} respectively, DAFs and VDAFs may impose
```

across multiple batches. It is RECOMMENDED that the nonce generated by the
Client be used by the Aggregators for replay protection.

## Requirements for the Aggregation Parameters

As described in {{sec-daf-validity-scopes}} and {{sec-vdaf-validity-scopes}}
respectively, DAFs and VDAFs may impose restrictions on the re-use of input
shares. This is to ensure that correlated randomness provided by the Client
through the input share is not used more than once, which might compromise
confidentiality of the Client's measurements.

> Protocols that make use of VDAFs therefore MUST ensure to call `Vdaf.is_valid`

Reviewer 2: nit: rephrasing
```suggestion
Protocols that make use of VDAFs therefore MUST call `Vdaf.is_valid`
```
